### PR TITLE
feat(ci): Automatically label PRs with merge conflicts

### PR DESCRIPTION
**Pull request by tibetiroka**
**CI/CD/Testing**

This PR adds a new workflow that adds/removes a special label from PRs with merge conflicts.

## Summary
Automatically adds a `has conflicts` label to PRs on conflicting pushes to master, and removes it once the conflicts are resolved. This helps to notify the PR authors of merge conflicts, who otherwise might not notice them.

We could also use the action to post a comment on the PR, but I opted not to do that.

## Testing Done
I can't test it, because github doesn't run workflows when there are merge conflicts, so it only works from the master branch.

### CODE REVIEW

**Comment by Hurleveur** [comment]
Honestly I wouldn't mind if this posted a comment on the branch, the issue is it would notify everyone instead of just the author. Can we somehow make it send a notif to the author only?

**Comment by tibetiroka** [comment]
I don't think github notifications work that way.

File: .github/workflows/label_merge_conflicts.yml
Comment (warp-core) [inline .github/workflows/label_merge_conflicts.yml:18]: When triggered by `pull_request_target`, this token has write access. https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target
We should probably use a `permissions` node in this job to only grant the permissions that it actually needs (modifying pull request labels).

I don't know what permissions if any it will have when running on a push to master; another reason to specify exactly what permissions are being granted.

Comment (tibetiroka) [reply .github/workflows/label_merge_conflicts.yml:22]: I think the most granular I can get is `pull-requests: write`, as that's what enables editing labels. The only more fine-grained option I could find is with granular PATs, but that's not really an option here.

**Comment by tibetiroka** [comment]
Cool. It runs now, but doesn't do anything as the `has conflicts` label doesn't exist yet. Creating that label might make this PR actually touch other PRs in the repo, which has me slightly worried... it's probably fine.

File: .github/workflows/label_merge_conflicts.yml
Comment (tibetiroka) [reply .github/workflows/label_merge_conflicts.yml:23]: https://docs.github.com/en/rest/issues/labels?apiVersion=2022-11-28#add-labels-to-an-issue

It needs `write` to both issues and pull requests, apparently.

Comment (warp-core) [reply .github/workflows/label_merge_conflicts.yml:23]: https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-pull-requests

https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-issues

Lists of what those permissions will allow.
I think explicitly:
```
issues: write
pull_requests: write
```
Is a lot better than whatever the token currently grants.

File: .github/workflows/label_merge_conflicts.yml
Comment (mOctave) [inline .github/workflows/label_merge_conflicts.yml:6]: Should this just be limited to a push to master? It's probably fine, but if in the future we switched to having multiple branches, it might cause problems.

File: .github/workflows/label_merge_conflicts.yml
Comment (mOctave) [inline .github/workflows/label_merge_conflicts.yml:1]:
```suggestion
name: Manage Labels
```
to match all the other workflows.

File: .github/workflows/label_merge_conflicts.yml
Comment (mOctave) [inline .github/workflows/label_merge_conflicts.yml:8]: Any reason not to run it on PR open as well, since people might PR something with merge conflicts in it already?

**Review by mOctave — APPROVED** [review]
The action version number and parameters look right, and I can't see any other major issues, just a couple of questions and one capitalization change.

File: .github/workflows/label_merge_conflicts.yml
Comment (tibetiroka) [reply .github/workflows/label_merge_conflicts.yml:6]: There are other workflows using these restrictions, so it's probably fine.

File: .github/workflows/label_merge_conflicts.yml
Comment (tibetiroka) [reply .github/workflows/label_merge_conflicts.yml:8]: If they already have merge conflicts, workflows won't run.

Comment (mOctave) [reply .github/workflows/label_merge_conflicts.yml:8]: Ah okay.

**Comment by TomGoodIdea** [comment]
What's the status of this PR?

**Comment by warp-core** [comment]
We want to test it somewhere other than here before merging it.

It also looks like it shouldn't work right now because it is running on the `pull_request` event which, as I understand it, will not be passed the token that can update labels on ORs.

**Comment by tibetiroka** [comment]
Actually, I'm just going to close it until we find a better way of managing workflow permissions.

I still think the `has conflicts` label should be created, though.

**Pull request closed without merging**
